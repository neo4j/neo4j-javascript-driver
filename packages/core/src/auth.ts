/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @property {function(username: string, password: string, realm: ?string)} basic the function to create a
 * basic authentication token.
 * @property {function(base64EncodedTicket: string)} kerberos the function to create a Kerberos authentication token.
 * Accepts a single string argument - base64 encoded Kerberos ticket.
 * @property {function(base64EncodedTicket: string)} bearer the function to create a Bearer authentication token.
 * Accepts a single string argument - base64 encoded Bearer ticket.
 * @property {function(principal: string, credentials: string, realm: string, scheme: string, parameters: ?object)} custom
 * the function to create a custom authentication token.
 */
 const auth = {
  basic: (username: string, password: string, realm?: string) => {
    if (realm) {
      return {
        scheme: 'basic',
        principal: username,
        credentials: password,
        realm: realm
      }
    } else {
      return { scheme: 'basic', principal: username, credentials: password }
    }
  },
  kerberos: (base64EncodedTicket: string) => {
    return {
      scheme: 'kerberos',
      principal: '', // This empty string is required for backwards compatibility.
      credentials: base64EncodedTicket
    }
  },
  bearer: (base64EncodedToken: string) => {
    return {
      scheme: 'bearer',
      credentials: base64EncodedToken
    }
  },
  custom: (
    principal: string,
    credentials: string,
    realm: string,
    scheme: string,
    parameters?: string
  ) => {
    if (parameters) {
      return {
        scheme: scheme,
        principal: principal,
        credentials: credentials,
        realm: realm,
        parameters: parameters
      }
    } else {
      return {
        scheme: scheme,
        principal: principal,
        credentials: credentials,
        realm: realm
      }
    }
  }
}

export default auth