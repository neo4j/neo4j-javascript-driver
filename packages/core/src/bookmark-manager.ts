/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export default interface BookmarkManager {
  /**
  * Method called when the bookmarks get update because of a given event
  *
  * @param database The database which the bookmarks belongs to
  * @param previousBookmarks The bookmarks used during the session creation
  * @param newBookmarks The new bookmarks resolved at the end of the session.
  * @returns {void}
  */
  updateBookmarks: (database: string, previousBookmarks: string[], newBookmarks: string[]) => void

  /**
   * Method called by the driver to get the bookmark for one specific database
   *
   * @param database The database which the bookmarks belongs to
   * @returns {string[]} The set of bookmarks
   */
  getBookmarks: (database: string) => string[]

  /**
   * Method called by the driver for getting all the bookmarks
   *
   * @param mustIncludedDatabases The database which must be included in the result even if they don't have be initialized yet.
   * @returns {string[]} The set of bookmarks
   */
  getAllBookmarks: (mustIncludedDatabases: string[]) => string[]
}

export interface BookmarkManagerConfig {
  initialBookmarks?: Map<string, string[]>
  bookmarkSupplier?: (database: string) => string[]
  notifyBookmarks?: (database: string, bookmarks: string[]) => void
}

export function bookmarkManager (config: BookmarkManagerConfig = {}): BookmarkManager {
  const initialBookmarks = new Map<string, Set<string>>()

  config.initialBookmarks?.forEach((v, k) => initialBookmarks.set(k, new Set(v)))

  return new Neo4jBookmarkManager(
    initialBookmarks,
    config.bookmarkSupplier,
    config.notifyBookmarks
  )
}

class Neo4jBookmarkManager implements BookmarkManager {
  constructor (
    private readonly _bookmarksPerDb: Map<string, Set<string>>,
    private readonly _bookmarkSupplier?: (database: string) => string[],
    private readonly _notifyBookmarks?: (database: string, bookmark: string[]) => void
  ) {

  }

  updateBookmarks (database: string, previousBookmarks: string[], newBookmarks: string[]): void {
    const bookmarks = this._getOrInitializeBookmarks(database)
    previousBookmarks.forEach(bm => bookmarks.delete(bm))
    newBookmarks.forEach(bm => bookmarks.add(bm))

    if (typeof this._notifyBookmarks === 'function') {
      this._notifyBookmarks(database, [...bookmarks])
    }
  }

  private _getOrInitializeBookmarks (database: string): Set<string> {
    let maybeBookmarks = this._bookmarksPerDb.get(database)
    if (maybeBookmarks == null) {
      maybeBookmarks = new Set()
      this._bookmarksPerDb.set(database, maybeBookmarks)
    }
    return maybeBookmarks
  }

  getBookmarks (database: string): string[] {
    const bookmarks = this._bookmarksPerDb.get(database) ?? []

    if (typeof this._bookmarkSupplier === 'function') {
      const suppliedBookmarks = this._bookmarkSupplier(database) ?? []
      return [...bookmarks, ...suppliedBookmarks]
    }

    return [...bookmarks]
  }

  getAllBookmarks (mustIncludedDatabases: string[]): string[] {
    const bookmarks = []
    const databases = new Set([...this._bookmarksPerDb.keys(), ...mustIncludedDatabases])

    for (const database of databases) {
      bookmarks.push(...this.getBookmarks(database))
    }

    return bookmarks
  }
}
