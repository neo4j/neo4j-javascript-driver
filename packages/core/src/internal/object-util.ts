/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const __isBrokenObject__ = '__isBrokenObject__'
// eslint-disable-next-line @typescript-eslint/naming-convention
const __reason__ = '__reason__'

/**
 * Creates a object on which all method calls will throw the given error
 *
 * @param {Error} error The error
 * @param {any} object The object. Default: {}
 * @returns {any} A broken object
 */
function createBrokenObject<T extends object> (error: Error, object: any = {}): T {
  const fail: <T>() => T = () => {
    throw error
  }

  return new Proxy(object, {
    get: (_: T, p: string | Symbol): any => {
      if (p === __isBrokenObject__) {
        return true
      } else if (p === __reason__) {
        return error
      } else if (p === 'toJSON') {
        return undefined
      }
      fail()
    },
    set: fail,
    apply: fail,
    construct: fail,
    defineProperty: fail,
    deleteProperty: fail,
    getOwnPropertyDescriptor: fail,
    getPrototypeOf: fail,
    has: fail,
    isExtensible: fail,
    ownKeys: fail,
    preventExtensions: fail,
    setPrototypeOf: fail
  })
}

/**
 * Verifies if it is a Broken Object
 * @param {any} object The object
 * @returns {boolean} If it was created with createBrokenObject
 */
function isBrokenObject (object: any): boolean {
  return object !== null && typeof object === 'object' && object[__isBrokenObject__] === true
}

/**
 * Returns if the reason the object is broken.
 *
 * This method should only be called with instances create with {@link createBrokenObject}
 *
 * @param {any} object The object
 * @returns {Error} The reason the object is broken
 */
function getBrokenObjectReason (object: any): Error {
  return object[__reason__]
}

export {
  createBrokenObject,
  isBrokenObject,
  getBrokenObjectReason
}
