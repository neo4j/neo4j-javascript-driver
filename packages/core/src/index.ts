/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  newError,
  Neo4jError,
  PROTOCOL_ERROR,
  SERVICE_UNAVAILABLE,
  SESSION_EXPIRED
} from './error'
import Integer, { int, isInt, inSafeRange, toNumber, toString } from './integer'
import {
  Date,
  DateTime,
  Duration,
  isDate,
  isDateTime,
  isDuration,
  isLocalDateTime,
  isLocalTime,
  isTime,
  LocalDateTime,
  LocalTime,
  Time
} from './temporal-types'
import {
  StandardDate,
  NumberOrInteger,
  Node,
  isNode,
  Relationship,
  isRelationship,
  UnboundRelationship,
  isUnboundRelationship,
  Path,
  isPath,
  PathSegment,
  isPathSegment
} from './graph-types'
import Record from './record'
import { isPoint, Point } from './spatial-types'
import ResultSummary, {
  queryType,
  ServerInfo,
  Notification,
  NotificationPosition,
  Plan,
  ProfiledPlan,
  QueryStatistics,
  Stats
} from './result-summary'
import Result, { QueryResult, ResultObserver } from './result'
import ConnectionProvider from './connection-provider'
import Connection from './connection'
import Transaction from './transaction'
import Session, { TransactionConfig } from './session'
import Driver, * as driver from './driver'
import auth from './auth'
import * as types from './types'
import * as json from './json'
import * as internal from './internal' // todo: removed afterwards

/**
 * Object containing string constants representing predefined {@link Neo4jError} codes.
 */
const error = {
  SERVICE_UNAVAILABLE,
  SESSION_EXPIRED,
  PROTOCOL_ERROR
}

/**
 * @private
 */
const forExport = {
  newError,
  Neo4jError,
  error,
  Integer,
  int,
  isInt,
  inSafeRange,
  toNumber,
  toString,
  internal,
  isPoint,
  Point,
  Date,
  DateTime,
  Duration,
  isDate,
  isDateTime,
  isDuration,
  isLocalDateTime,
  isLocalTime,
  isTime,
  LocalDateTime,
  LocalTime,
  Time,
  Node,
  isNode,
  Relationship,
  isRelationship,
  UnboundRelationship,
  isUnboundRelationship,
  Path,
  isPath,
  PathSegment,
  isPathSegment,
  Record,
  ResultSummary,
  queryType,
  ServerInfo,
  Notification,
  Plan,
  ProfiledPlan,
  QueryStatistics,
  Stats,
  Result,
  Transaction,
  Session,
  Driver,
  Connection,
  types,
  driver,
  json,
  auth
}

export {
  newError,
  Neo4jError,
  error,
  Integer,
  int,
  isInt,
  inSafeRange,
  toNumber,
  toString,
  internal,
  isPoint,
  Point,
  Date,
  DateTime,
  Duration,
  isDate,
  isDateTime,
  isDuration,
  isLocalDateTime,
  isLocalTime,
  isTime,
  LocalDateTime,
  LocalTime,
  Time,
  Node,
  isNode,
  Relationship,
  isRelationship,
  UnboundRelationship,
  isUnboundRelationship,
  Path,
  isPath,
  PathSegment,
  isPathSegment,
  Record,
  ResultSummary,
  queryType,
  ServerInfo,
  Notification,
  Plan,
  ProfiledPlan,
  QueryStatistics,
  Stats,
  Result,
  ConnectionProvider,
  Connection,
  Transaction,
  Session,
  Driver,
  types,
  driver,
  json,
  auth
}

export type {
  StandardDate,
  NumberOrInteger,
  NotificationPosition,
  QueryResult,
  ResultObserver,
  TransactionConfig,
};

export default forExport
