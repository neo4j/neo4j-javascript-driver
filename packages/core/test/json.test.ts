/**
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { json, newError } from '../src'
import { createBrokenObject } from '../src/internal/object-util'

describe('json', () => {
  describe('.stringify', () => {
    it('should handle objects created with createBrokenObject', () => {
      const reason = newError('some error')
      const broken = createBrokenObject(reason, { })

      expect(json.stringify(broken)).toMatchSnapshot()
    })

    it('should handle objects created with createBrokenObject in list', () => {
      const reason = newError('some error')
      const broken = createBrokenObject(reason, { })

      expect(json.stringify([broken])).toMatchSnapshot()
    })

    it('should handle objects created with createBrokenObject inside other object', () => {
      const reason = newError('some error')
      const broken = createBrokenObject(reason, { })

      expect(json.stringify({
        number: 1,
        broken
      })).toMatchSnapshot()
    })
  })
})
