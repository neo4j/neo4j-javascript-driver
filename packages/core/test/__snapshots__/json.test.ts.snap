// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json .stringify should handle Date 1`] = `"{"year":1999,"month":4,"day":12}"`;

exports[`json .stringify should handle Date in list 1`] = `"[{"year":1999,"month":4,"day":12}]"`;

exports[`json .stringify should handle Date in object 1`] = `"{"key":{"year":1999,"month":4,"day":12}}"`;

exports[`json .stringify should handle DateTime 1`] = `"{"year":2024,"month":6,"day":13,"hour":10,"minute":0,"second":30,"nanosecond":134,"timeZoneOffsetSeconds":-3600,"timeZoneId":"Europe/Berlin"}"`;

exports[`json .stringify should handle DateTime in list 1`] = `"[{"year":2024,"month":6,"day":13,"hour":10,"minute":0,"second":30,"nanosecond":134,"timeZoneOffsetSeconds":-3600,"timeZoneId":"Europe/Berlin"}]"`;

exports[`json .stringify should handle DateTime in object 1`] = `"{"key":{"year":2024,"month":6,"day":13,"hour":10,"minute":0,"second":30,"nanosecond":134,"timeZoneOffsetSeconds":-3600,"timeZoneId":"Europe/Berlin"}}"`;

exports[`json .stringify should handle Duration 1`] = `"{"months":10,"days":2,"seconds":{"low":35,"high":0},"nanoseconds":{"low":100,"high":0}}"`;

exports[`json .stringify should handle Duration in list 1`] = `"[{"months":10,"days":2,"seconds":{"low":35,"high":0},"nanoseconds":{"low":100,"high":0}}]"`;

exports[`json .stringify should handle Duration in object 1`] = `"{"key":{"months":10,"days":2,"seconds":{"low":35,"high":0},"nanoseconds":{"low":100,"high":0}}}"`;

exports[`json .stringify should handle Integer 1`] = `"{"low":5,"high":0}"`;

exports[`json .stringify should handle Integer in list 1`] = `"[{"low":5,"high":0}]"`;

exports[`json .stringify should handle Integer in object 1`] = `"{"key":{"low":5,"high":0}}"`;

exports[`json .stringify should handle LocalDateTime 1`] = `"{"year":1999,"month":4,"day":28,"hour":12,"minute":40,"second":12,"nanosecond":301}"`;

exports[`json .stringify should handle LocalDateTime in list 1`] = `"[{"year":1999,"month":4,"day":28,"hour":12,"minute":40,"second":12,"nanosecond":301}]"`;

exports[`json .stringify should handle LocalDateTime in object 1`] = `"{"key":{"year":1999,"month":4,"day":28,"hour":12,"minute":40,"second":12,"nanosecond":301}}"`;

exports[`json .stringify should handle LocalTime 1`] = `"{"hour":2,"minute":30,"second":25,"nanosecond":150}"`;

exports[`json .stringify should handle LocalTime in list 1`] = `"[{"hour":2,"minute":30,"second":25,"nanosecond":150}]"`;

exports[`json .stringify should handle LocalTime in object 1`] = `"{"key":{"hour":2,"minute":30,"second":25,"nanosecond":150}}"`;

exports[`json .stringify should handle Node 1`] = `"{"identity":1,"labels":["Person"],"properties":{"name":"Mr. Bauer"},"elementId":"myId"}"`;

exports[`json .stringify should handle Node in list 1`] = `"[{"identity":1,"labels":["Person"],"properties":{"name":"Mr. Bauer"},"elementId":"myId"}]"`;

exports[`json .stringify should handle Node in object 1`] = `"{"key":{"identity":1,"labels":["Person"],"properties":{"name":"Mr. Bauer"},"elementId":"myId"}}"`;

exports[`json .stringify should handle Path 1`] = `"{"start":{"identity":1,"labels":["Person"],"properties":{"name":"Antonio"},"elementId":"antonioId"},"end":{"identity":2,"labels":["Person"],"properties":{"name":"Mr. Bauer"},"elementId":"mrBauerId"},"segments":[{"start":{"identity":1,"labels":["Person"],"properties":{"name":"Antonio"},"elementId":"antonioId"},"relationship":{"identity":3,"start":1,"end":2,"type":"PLAY_FOOTBALL","properties":{"since":1897},"elementId":"relId","startNodeElementId":"antonioId","endNodeElementId":"mrBauerId"},"end":{"identity":2,"labels":["Person"],"properties":{"name":"Mr. Bauer"},"elementId":"mrBauerId"}}],"length":1}"`;

exports[`json .stringify should handle Path in list 1`] = `"[{"start":{"identity":1,"labels":["Person"],"properties":{"name":"Antonio"},"elementId":"antonioId"},"end":{"identity":2,"labels":["Person"],"properties":{"name":"Mr. Bauer"},"elementId":"mrBauerId"},"segments":[{"start":{"identity":1,"labels":["Person"],"properties":{"name":"Antonio"},"elementId":"antonioId"},"relationship":{"identity":3,"start":1,"end":2,"type":"PLAY_FOOTBALL","properties":{"since":1897},"elementId":"relId","startNodeElementId":"antonioId","endNodeElementId":"mrBauerId"},"end":{"identity":2,"labels":["Person"],"properties":{"name":"Mr. Bauer"},"elementId":"mrBauerId"}}],"length":1}]"`;

exports[`json .stringify should handle Path in object 1`] = `"{"key":{"start":{"identity":1,"labels":["Person"],"properties":{"name":"Antonio"},"elementId":"antonioId"},"end":{"identity":2,"labels":["Person"],"properties":{"name":"Mr. Bauer"},"elementId":"mrBauerId"},"segments":[{"start":{"identity":1,"labels":["Person"],"properties":{"name":"Antonio"},"elementId":"antonioId"},"relationship":{"identity":3,"start":1,"end":2,"type":"PLAY_FOOTBALL","properties":{"since":1897},"elementId":"relId","startNodeElementId":"antonioId","endNodeElementId":"mrBauerId"},"end":{"identity":2,"labels":["Person"],"properties":{"name":"Mr. Bauer"},"elementId":"mrBauerId"}}],"length":1}}"`;

exports[`json .stringify should handle Point 1`] = `"{"srid":4979,"x":1,"y":2,"z":3}"`;

exports[`json .stringify should handle Point in list 1`] = `"[{"srid":4979,"x":1,"y":2,"z":3}]"`;

exports[`json .stringify should handle Point in object 1`] = `"{"key":{"srid":4979,"x":1,"y":2,"z":3}}"`;

exports[`json .stringify should handle Relationship 1`] = `"{"identity":1,"start":2,"end":3,"type":"FRIENDSHIP","properties":{"started":1999},"elementId":"myId","startNodeElementId":"startId","endNodeElementId":"endId"}"`;

exports[`json .stringify should handle Relationship in list 1`] = `"[{"identity":1,"start":2,"end":3,"type":"FRIENDSHIP","properties":{"started":1999},"elementId":"myId","startNodeElementId":"startId","endNodeElementId":"endId"}]"`;

exports[`json .stringify should handle Relationship in object 1`] = `"{"key":{"identity":1,"start":2,"end":3,"type":"FRIENDSHIP","properties":{"started":1999},"elementId":"myId","startNodeElementId":"startId","endNodeElementId":"endId"}}"`;

exports[`json .stringify should handle Time 1`] = `"{"hour":12,"minute":50,"second":23,"nanosecond":300,"timeZoneOffsetSeconds":3600}"`;

exports[`json .stringify should handle Time in list 1`] = `"[{"hour":12,"minute":50,"second":23,"nanosecond":300,"timeZoneOffsetSeconds":3600}]"`;

exports[`json .stringify should handle Time in object 1`] = `"{"key":{"hour":12,"minute":50,"second":23,"nanosecond":300,"timeZoneOffsetSeconds":3600}}"`;

exports[`json .stringify should handle UnboundRelationship 1`] = `"{"identity":1,"type":"ALONE","properties":{"since":2001},"elementId":"myId"}"`;

exports[`json .stringify should handle UnboundRelationship in list 1`] = `"[{"identity":1,"type":"ALONE","properties":{"since":2001},"elementId":"myId"}]"`;

exports[`json .stringify should handle UnboundRelationship in object 1`] = `"{"key":{"identity":1,"type":"ALONE","properties":{"since":2001},"elementId":"myId"}}"`;

exports[`json .stringify should handle bigint 1`] = `""3n""`;

exports[`json .stringify should handle bigint in list 1`] = `"["3n"]"`;

exports[`json .stringify should handle bigint in object 1`] = `"{"key":"3n"}"`;

exports[`json .stringify should handle list 1`] = `"["1",2,{"tres":3}]"`;

exports[`json .stringify should handle list in list 1`] = `"[["1",2,{"tres":3}]]"`;

exports[`json .stringify should handle list in object 1`] = `"{"key":["1",2,{"tres":3}]}"`;

exports[`json .stringify should handle number 1`] = `"2"`;

exports[`json .stringify should handle number in list 1`] = `"[2]"`;

exports[`json .stringify should handle number in object 1`] = `"{"key":2}"`;

exports[`json .stringify should handle object 1`] = `"{"identity":123,"labels":["a"],"properties":{"key":"value"},"elementId":"abc"}"`;

exports[`json .stringify should handle object in list 1`] = `"[{"identity":123,"labels":["a"],"properties":{"key":"value"},"elementId":"abc"}]"`;

exports[`json .stringify should handle object in object 1`] = `"{"key":{"identity":123,"labels":["a"],"properties":{"key":"value"},"elementId":"abc"}}"`;

exports[`json .stringify should handle object with custom toString 1`] = `"{"identity":"1"}"`;

exports[`json .stringify should handle object with custom toString in list 1`] = `"[{"identity":"1"}]"`;

exports[`json .stringify should handle object with custom toString in object 1`] = `"{"key":{"identity":"1"}}"`;

exports[`json .stringify should handle objects created with createBrokenObject 1`] = `"{"__isBrokenObject__":true,"__reason__":{"code":"N/A","name":"Neo4jError","retriable":false}}"`;

exports[`json .stringify should handle objects created with createBrokenObject in list 1`] = `"[{"__isBrokenObject__":true,"__reason__":{"code":"N/A","name":"Neo4jError","retriable":false}}]"`;

exports[`json .stringify should handle objects created with createBrokenObject inside other object 1`] = `"{"number":1,"broken":{"__isBrokenObject__":true,"__reason__":{"code":"N/A","name":"Neo4jError","retriable":false}}}"`;

exports[`json .stringify should handle string 1`] = `""my string""`;

exports[`json .stringify should handle string in list 1`] = `"["my string"]"`;

exports[`json .stringify should handle string in object 1`] = `"{"key":"my string"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle Date 1`] = `""1999-04-12""`;

exports[`json .stringify when opts.useCustomToString=true should handle Date in list 1`] = `"["1999-04-12"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle Date in object 1`] = `"{"key":"1999-04-12"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle DateTime 1`] = `""2024-06-13T10:00:30.000000134-01:00[Europe/Berlin]""`;

exports[`json .stringify when opts.useCustomToString=true should handle DateTime in list 1`] = `"["2024-06-13T10:00:30.000000134-01:00[Europe/Berlin]"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle DateTime in object 1`] = `"{"key":"2024-06-13T10:00:30.000000134-01:00[Europe/Berlin]"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle Duration 1`] = `""P10M2DT35.000000100S""`;

exports[`json .stringify when opts.useCustomToString=true should handle Duration in list 1`] = `"["P10M2DT35.000000100S"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle Duration in object 1`] = `"{"key":"P10M2DT35.000000100S"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle Integer 1`] = `""5""`;

exports[`json .stringify when opts.useCustomToString=true should handle Integer in list 1`] = `"["5"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle Integer in object 1`] = `"{"key":"5"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle LocalDateTime 1`] = `""1999-04-28T12:40:12.000000301""`;

exports[`json .stringify when opts.useCustomToString=true should handle LocalDateTime in list 1`] = `"["1999-04-28T12:40:12.000000301"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle LocalDateTime in object 1`] = `"{"key":"1999-04-28T12:40:12.000000301"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle LocalTime 1`] = `""02:30:25.000000150""`;

exports[`json .stringify when opts.useCustomToString=true should handle LocalTime in list 1`] = `"["02:30:25.000000150"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle LocalTime in object 1`] = `"{"key":"02:30:25.000000150"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle Node 1`] = `""(myId:Person {name:\\"Mr. Bauer\\"})""`;

exports[`json .stringify when opts.useCustomToString=true should handle Node in list 1`] = `"["(myId:Person {name:\\"Mr. Bauer\\"})"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle Node in object 1`] = `"{"key":"(myId:Person {name:\\"Mr. Bauer\\"})"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle Path 1`] = `"{"start":"(antonioId:Person {name:\\"Antonio\\"})","end":"(mrBauerId:Person {name:\\"Mr. Bauer\\"})","segments":[{"start":"(antonioId:Person {name:\\"Antonio\\"})","relationship":"(antonioId)-[:PLAY_FOOTBALL {since:1897}]->(mrBauerId)","end":"(mrBauerId:Person {name:\\"Mr. Bauer\\"})"}],"length":1}"`;

exports[`json .stringify when opts.useCustomToString=true should handle Path in list 1`] = `"[{"start":"(antonioId:Person {name:\\"Antonio\\"})","end":"(mrBauerId:Person {name:\\"Mr. Bauer\\"})","segments":[{"start":"(antonioId:Person {name:\\"Antonio\\"})","relationship":"(antonioId)-[:PLAY_FOOTBALL {since:1897}]->(mrBauerId)","end":"(mrBauerId:Person {name:\\"Mr. Bauer\\"})"}],"length":1}]"`;

exports[`json .stringify when opts.useCustomToString=true should handle Path in object 1`] = `"{"key":{"start":"(antonioId:Person {name:\\"Antonio\\"})","end":"(mrBauerId:Person {name:\\"Mr. Bauer\\"})","segments":[{"start":"(antonioId:Person {name:\\"Antonio\\"})","relationship":"(antonioId)-[:PLAY_FOOTBALL {since:1897}]->(mrBauerId)","end":"(mrBauerId:Person {name:\\"Mr. Bauer\\"})"}],"length":1}}"`;

exports[`json .stringify when opts.useCustomToString=true should handle Point 1`] = `""Point{srid=4979.0, x=1.0, y=2.0, z=3.0}""`;

exports[`json .stringify when opts.useCustomToString=true should handle Point in list 1`] = `"["Point{srid=4979.0, x=1.0, y=2.0, z=3.0}"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle Point in object 1`] = `"{"key":"Point{srid=4979.0, x=1.0, y=2.0, z=3.0}"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle Relationship 1`] = `""(startId)-[:FRIENDSHIP {started:1999}]->(endId)""`;

exports[`json .stringify when opts.useCustomToString=true should handle Relationship in list 1`] = `"["(startId)-[:FRIENDSHIP {started:1999}]->(endId)"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle Relationship in object 1`] = `"{"key":"(startId)-[:FRIENDSHIP {started:1999}]->(endId)"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle Time 1`] = `""12:50:23.000000300+01:00""`;

exports[`json .stringify when opts.useCustomToString=true should handle Time in list 1`] = `"["12:50:23.000000300+01:00"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle Time in object 1`] = `"{"key":"12:50:23.000000300+01:00"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle UnboundRelationship 1`] = `""-[:ALONE {since:2001}]->""`;

exports[`json .stringify when opts.useCustomToString=true should handle UnboundRelationship in list 1`] = `"["-[:ALONE {since:2001}]->"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle UnboundRelationship in object 1`] = `"{"key":"-[:ALONE {since:2001}]->"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle bigint 1`] = `""3n""`;

exports[`json .stringify when opts.useCustomToString=true should handle bigint in list 1`] = `"["3n"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle bigint in object 1`] = `"{"key":"3n"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle list 1`] = `"["1",2,{"tres":3}]"`;

exports[`json .stringify when opts.useCustomToString=true should handle list in list 1`] = `"[["1",2,{"tres":3}]]"`;

exports[`json .stringify when opts.useCustomToString=true should handle list in object 1`] = `"{"key":["1",2,{"tres":3}]}"`;

exports[`json .stringify when opts.useCustomToString=true should handle number 1`] = `"2"`;

exports[`json .stringify when opts.useCustomToString=true should handle number in list 1`] = `"[2]"`;

exports[`json .stringify when opts.useCustomToString=true should handle number in object 1`] = `"{"key":2}"`;

exports[`json .stringify when opts.useCustomToString=true should handle object 1`] = `"{"identity":123,"labels":["a"],"properties":{"key":"value"},"elementId":"abc"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle object in list 1`] = `"[{"identity":123,"labels":["a"],"properties":{"key":"value"},"elementId":"abc"}]"`;

exports[`json .stringify when opts.useCustomToString=true should handle object in object 1`] = `"{"key":{"identity":123,"labels":["a"],"properties":{"key":"value"},"elementId":"abc"}}"`;

exports[`json .stringify when opts.useCustomToString=true should handle object with custom toString 1`] = `""My identity is One""`;

exports[`json .stringify when opts.useCustomToString=true should handle object with custom toString in list 1`] = `"["My identity is One"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle object with custom toString in object 1`] = `"{"key":"My identity is One"}"`;

exports[`json .stringify when opts.useCustomToString=true should handle string 1`] = `""my string""`;

exports[`json .stringify when opts.useCustomToString=true should handle string in list 1`] = `"["my string"]"`;

exports[`json .stringify when opts.useCustomToString=true should handle string in object 1`] = `"{"key":"my string"}"`;
